In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', 50)
sns.set_theme(style='whitegrid')

# this is the largest dataset - 2 million rows of daily chart data
# across 73 countries from 2023 to 2025
# it is very large so it may take a moment to load
df_charts = pd.read_csv('/Users/kultumlhabaik/Documents/data-mining-final-project/data/raw/universal_top_spotify_songs.csv')

print('Rows:', len(df_charts))
print('Columns:', df_charts.shape[1])
df_charts.head()

Rows: 2110316
Columns: 25


,spotify_id,name,artists,daily_rank,daily_movement,weekly_movement,country,snapshot_date,popularity,is_explicit,duration_ms,album_name,album_release_date,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature
0,2RkZ5LkEzeHGRsmDqKwmaJ,Ordinary,Alex Warren,1,1,0,NaN,2025-06-11,95,False,186964,"You'll Be Alright, Kid (Chapter 1)",2024-09-26,0.368,0.694,2,-6.141,1,0.0600,0.704000,0.000007,0.0550,0.391,168.115,3
1,42UBPzRMh5yyz0EDPr6fr1,Manchild,Sabrina Carpenter,2,-1,48,NaN,2025-06-11,89,True,213645,Manchild,2025-06-05,0.731,0.685,7,-5.087,1,0.0572,0.122000,0.000000,0.3170,0.811,123.010,4
2,0FTmksd2dxiE5e3rWyJXs6,back to friends,sombr,3,0,1,NaN,2025-06-11,98,False,199032,back to friends,2024-12-27,0.436,0.723,1,-2.291,1,0.0301,0.000094,0.000088,0.0929,0.235,92.855,4
3,7so0lgd0zP2Sbgs2d7a1SZ,Die With A Smile,"Lady Gaga, Bruno Mars",4,0,-1,NaN,2025-06-11,91,False,251667,MAYHEM,2025-03-07,0.519,0.601,6,-7.727,0,0.0317,0.289000,0.000000,0.1260,0.498,157.964,3
4,6dOtVTDdiauQNBQEDOtlAB,BIRDS OF A FEATHER,Billie Eilish,5,1,0,NaN,2025-06-11,100,False,210373,HIT ME HARD AND SOFT,2024-05-17,0.747,0.507,2,-10.171,1,0.0358,0.200000,0.060800,0.1170,0.438,104.978,4


In [5]:
# check date range, countries, and missing values
print('Date Range')
print('Earliest date:', df_charts['snapshot_date'].min())
print('Latest date:  ', df_charts['snapshot_date'].max())

print('\nCountries')
print('Number of countries:', df_charts['country'].nunique())
print('Global chart rather than any specific country: ', df_charts['country'].isnull().sum())

print('\nMissing Values')
missing = df_charts.isnull().sum()
print(missing[missing > 0])

Date Range
Earliest date: 2023-10-18
Latest date:   2025-06-11

Countries
Number of countries: 72
Global chart rather than any specific country:  28908

Missing Values
name                     30
artists                  29
country               28908
album_name              822
album_release_date      659
dtype: int64


## universal_top_spotify_songs.csv — Decision

This dataset covers daily Spotify charts from 2023-10-18 to 2025-06-11
across 72 countries with 2,110,316 rows.

It cannot be used in this project for one critical reason:
Our main dataset covers 1990-2020 and this dataset starts in 2023.
There overlap between them, so it cannot be used
to help our artist trajectory analysis.

Final decision: This dataset will not be used.